src/python/pants/base/cmd_line_spec_parser.py

```diff
@@ -72,13 +75,15 @@ def parse_spec(self, spec: str) -> Union[AddressSpec, FilesystemSpec]:
       spec_parts = spec.rsplit(':', 1)
       spec_path = self._normalize_spec_path(spec_parts[0])
       return SingleAddress(directory=spec_path, name=spec_parts[1])
-    if spec.startswith('!') or '*' in spec:
-      return FilesystemSpec(glob=spec)
+    if spec.startswith('!'):
+      return FilesystemIgnoreSpec(spec[1:])
+    if '*' in spec:
+      return FilesystemGlobSpec(spec)
```

Comment:
This is getting large enough that I think there might be room for a parser generator here. I'm not sure that's necessary now, but having a formalization of a language (tends to) make it much easier to debug and understand at a glance.